### ISSUE TITLE
Units interface package

### ISSUE THREAD

**Issue by tfiers** [PainterQubits/Unitful.jl#595]
`using Unitful` currently takes 1 to 3 seconds[^timing].

This is fine if you're planning on actually using physical units in your Julia session.[^precompile]
But it adds to the load time of downstream packages that offer _optional_ Unitful support.

My concrete motivation for writing this issue is that I want to add Unitful support to Distributions.jl:
- https://github.com/JuliaStats/Distributions.jl/issues/1413#issuecomment-1301249731

Doing that would require adding Unitful as a dependency to Distributions. But most Distributions.jl users will not use the new unitful distributions; and they would still get the added start-up latency.

That is why I propose to create a very **lightweight package that defines just the interface/API** for working with units in Julia. Something like `UnitfulCore`, or `UnitsInterface.jl`.

Downstream packages offering physical unit support would then load just that interface package.
(If the user wants to actually  _use_ the unit support, they have to explicitly be `using Unitful, DownstreamPkg`).

This pattern has been applied before, e.g. in StaticArrays.jl. From [its readme](https://github.com/JuliaArrays/StaticArrays.jl):
> Most of the primary array types exported by StaticArrays.jl are defined in the small interface package [StaticArraysCore.jl](https://github.com/JuliaArrays/StaticArraysCore.jl). This includes e.g., the definitions of the abstract type `StaticArray` and the concrete types `SArray`, `MArray`, and `SizedArray` (as well as their dimension-specific aliases). This enables downstream packages to implement new methods for these types without depending on (and hence loading) the entirety of StaticArrays.jl, and thereby to avoid incurring the full load-time of StaticArrays.jl (which is on the order of 0.6 s for StaticArrays.jl v1.4 on Julia v1.7).

---

So my question to the maintainers of Unitful.jl is: would you be open to adding such an interface package as a dependency of Unitful, and moving some core types and functions to that package (or having them subtype from the package) ?

<br>
<br>

[^timing]: `@time using Unitful` gives 1.0 seconds [on replit][replit], and between 1.6 and 2.6 seconds on my laptop (Julia 1.8).
[^precompile]:  Though a [`@precompile_all_calls`][snoop] block might be a good idea to add to src/ in any case.

[snoop]: https://timholy.github.io/SnoopCompile.jl/stable/snoop_pc/
[replit]: https://replit.com/@TomasFiers/Unitful-load-time?v=1

**Comment by sostock** [comment]
There have been several requests for such an interface package, so I think we should consider it. Unfortunately, I didn’t have much time in the past to really take part in the previous discussion:

* #359, #424 and the related PR #428: These are not about loading times but about better support for other unit systems: Unitful currently defines dimensions, units and constants based on the SI system. The proposed changes would move everything except these definitions into UnitfulBase. In that case, UnitfulBase would still contain almost all of the current Unitful code. Therefore, it differs from the idea presented here, for which UnitfulBase should be as small as possible.
* https://github.com/PainterQubits/Unitful.jl/issues/512#issuecomment-1006817148 proposes a light package just for dispatching on dimensions (e.g., defining methods like `foo(x::Length, y::Energy)`).

Based on that, I think we would need a  hierarchy of several packages to satisfy most users’ needs. Maybe we can base the discussion on the following proposal and refine it:
1. **UnitfulCore**: Defines the `AbstractQuantity`, `Dimensions` and `Units` types and some functions that are required to facilitate working with quantities. Note that the actual implementation of `Quantity` with all of its arithmetic and promotion rules is not included, and neither are `FreeUnits` etc.
3. **UnitfulBase**: Depends on UnitfulCore and defines the `Quantity` type and implements the actual conversion and arithmetic. The `u"…"` macro should probably be defined here, as well as the `@unit`, `@dimension` macros etc.
4. **UnitfulDimensions**: Depends only on UnitfulCore, not on UnitfulBase. Defines the dimensions that are currently defined in Unitful (e.g. `𝐋`) and “dispatch types” (`Length` and `LengthUnits`). Since the package does not depend on UnitfulBase, these will have to be defined by hand, i.e., without using `@dimension`. Or should we include `@dimension` in UnitfulCore instead?
5. **Unitful**: Depends on UnitfulBase and UnitfulDimensions. Defines the units and constants that are currently included in `src/pkgdefaults.jl` as well as the methods for handling `Period` types from the Dates stdlib.

@giordano @Gregstrq @aplavin @mileslucas @tfiers What do you think about such a package hierarchy?

Some other thoughts:
* Even if we don’t do this, we should look into reducing the loading time of this package.
* Some packages might not need to depend on Unitful to support quantities. Maybe we could add a page to the documentation with some hints on how to do that? There is an older Julia PR (https://github.com/JuliaLang/julia/pull/39852) which I think should be revived. It adds functionality to the Test stdlib that allows users to test whether their functions work with dimensionful quantities without assuming any interface for them.
* We could think about renaming
  ```
  UnitfulCore       → UnitfulBase
  UnitfulBase       → Unitful
  UnitfulDimensions → UnitfulSIDimensions
  Unitful           → UnitfulSI
  ```
  This would be breaking (because Unitful wouldn’t contain the default SI dimensions/units/constants anymore), but we wouldn’t have both UnitfulCore and UnitfulBase anymore, which might be less confusing. It also would emphasize that the SI system is not necessarily the default. If we do this, we could also think about removing the constants from UnitfulSI altogether and refer to the PhysicalConstants package for that.

**Comment by cstjean** [comment]
I'm going to throw in my unsolicited opinion, as someone who's done a `*Base` package, that managing multiple packages is a big hassle. CI, github permissions, PRs that must now be spread across multiple repos, package compatibility, user confusion ("which package do I need?"), documentation... I haven't reviewed all the issues linked above, but IMVHO, there must be some big benefits to justify the costs. Quadruply so to justify a 4-way split.

**Comment by giordano** [comment]
> `using Unitful` currently takes 1 to 3 seconds.

```console
% julia --project=/tmp --startup-file=no -q
julia> @time @eval using Unitful
  0.444807 seconds (924.10 k allocations: 64.710 MiB, 5.12% gc time)
```

I have a 7-year old laptop.

**Comment by tfiers** [comment]
Interesting.
I get
```
julia> @time using Unitful
  0.781682 seconds (943.66 k allocations: 63.208 MiB)
```
if I run it on a very modern desktop (different than the modern laptop used for the earlier measurements)

**Comment by giordano** [comment]
> Aha, it's indeed shorter in an empty env (which is weird, right?!)

Are you always in a fresh session?  Maybe  there is something invalidating Unitful code?  Also, do you use Revise?  I see you use Windows, that may account for some slowness (I/O is pretty terrible on Windows).

**Comment by tfiers** [comment]
See my last edit (sorry for post-editing my comment):
```
$ julia --startup-file=no -e "@time using Unitful"
  0.840768 seconds (938.31 k allocations: 65.985 MiB, 2.28% gc time)
```

**Comment by Gregstrq** [comment]
I would agree with @cstjean that making four way split is a bit too much. Also, I can't imagine someone using UnitfulCore without the arithmetics and convenience constructors.

I stand by my initial proposal to only split out pkgdefaults.
In this regards, the dimensions are a part of the unit system, so there is no actual reason to split them out from pkgdefaults.

**Comment by sostock** [comment]
> I can't imagine someone using UnitfulCore without the arithmetics and convenience constructors.

In my opinion, UnitfulCore is all that should be needed to implement something like https://github.com/JuliaStats/Distributions.jl/issues/1413#issuecomment-1301249731, which is the reason why this issue was opened.

**Comment by tfiers** [comment]
I'd like to add that I plan to experiment with an alternative units package too (one with runtime units instead of units in the typesystem -- the goal is not necessarily performance, but rather ease of use and small type-signatures in compound datatypes with units).

So it would be fantastic to have a kind of UnitsInterface package that can be implemented by different packages.

Ofc it's great if everyone's effort is concentrated on the same package instead of duplicated; but independent experimentation is worthwile imho too

**Comment by KristofferC** [comment]
"Interface" packages and "Core" packages should have no purpose now that "package extensions" are available (or at least will be available in 1.9).

**Comment by tfiers** [comment]
For import times, indeed.
They might still be useful for allowing different implementations, no?

**Comment by timholy** [comment]
> managing multiple packages is a big hassle

Agreed. IMO "subdir" packages, in which you still use a single repo and can thus coordinate changes across multiple packages, reduce most but not all of this hassle. You still have to think through what you want to release, set up your CI scripts appropriately, etc.

**Comment by MilesCranmer** [comment]
Nice work!!

And yes there is buy-in from me on the DQ side. Can we move it to a shared Julia org to lower the bus factor?

**Comment by MilesCranmer** [comment]
JuliaMath sounds good (not sure who to ask for this?). Or maybe JuliaPhysics since that's where Measurements.jl and PhysicalConstants.jl live.

> Mixed Unitful + DQ arithmetic and conversion makes most sense to put into DQ.

Sounds good to me. We can put it in the existing Unitful extension within DQ.

My initial suggestions for more content would be `AbstractQuantity{T} <: Number` (and perhaps with a `AbstractRealQuantity{T} <: Real`).

**Comment by giordano** [comment]
I sent you an invite to JuliaPhysics.  Years ago I also proposed to move this repo to the same org (#228), but it eventually was moved here and I think we're stuck with it, Andrew is the only admin of the repo and he doesn't seem to be very active on GitHub anymore.

**Comment by MilesCranmer** [comment]
<img width="1170" height="2532" alt="Image" src="https://github.com/user-attachments/assets/9b60c3c7-89f7-4a68-9fa8-30977bc7f50c" />

Cool I guess this is the way. And just to confirm, I would still effectively own the project, have push rights, etc., it’s just helps security and stability to have more trusted people as maintainers?

**Comment by giordano** [comment]
Correct.

**Comment by MilesCranmer** [comment]
Thanks. Done! https://github.com/JuliaPhysics/DynamicQuantities.jl
